src/CanvasRenderingContext2d.cc

```diff
@@ -292,6 +292,58 @@ Context2d::restorePath() {
   cairo_path_destroy(_path);
 }
 
+/*
+ * Create temporary surface for gradient or pattern transparency
+ */
+cairo_pattern_t*
+create_transparent_gradient(cairo_pattern_t *source, float alpha) {
+  double x0;
+  double y0;
+  double x1;
+  double y1;
+  double r0;
+  double r1;
+  int count;
+  int i;
+  double offset;
+  double r;
+  double g;
+  double b;
+  double a;
+  cairo_pattern_t *newGradient;
+  cairo_pattern_type_t type = cairo_pattern_get_type(source);
+  cairo_pattern_get_color_stop_count(source, &count);
+  if (type == 2) {
+    cairo_pattern_get_linear_points (source, &x0, &y0, &x1, &y1);
+    newGradient = cairo_pattern_create_linear(x0, y0, x1, y1);
+  }
+  if (type == 3) {
+    cairo_pattern_get_radial_circles(source, &x0, &y0, &r0, &x1, &y1, &r1);
+    newGradient = cairo_pattern_create_radial (x0, y0, r0, x1, y1, r1);
+  }
+  for ( i = 0; i < count; i++ ) {
+    cairo_pattern_get_color_stop_rgba(source, i, &offset, &r, &g, &b, &a);
+    cairo_pattern_add_color_stop_rgba(newGradient, offset, r, g, b, a * alpha);
+  }
+  return newGradient;
+}
+
+cairo_pattern_t*
+create_transparent_pattern(cairo_pattern_t *source, float alpha) {
+  cairo_surface_t *surface;
+  cairo_pattern_get_surface(source, &surface);
+  int width = cairo_image_surface_get_width(surface);
+  int height = cairo_image_surface_get_height(surface);
+  cairo_surface_t *mask_surface = cairo_image_surface_create(
```

Comment:
`*mask_surface` needs to be freed at some point. I *think* it's safe to call [`cairo_surface_destroy`](https://www.cairographics.org/manual/cairo-cairo-surface-t.html#cairo-surface-destroy) just after your `cairo_pattern_create_for_surface(mask_surface)` call, so you can do memory management within this function.

Reply:
I still don't see this being freed... calling `cairo_surface_destroy` decreases the ref count, but doesn't destroy it unless the ref count is 0. `cairo_pattern_create_for_surface` [increases the ref count on the surface](https://github.molgen.mpg.de/git-mirror/cairo/blob/747cab741cf63f30aa6bc2b787fc13e7e7a2b859/src/cairo-pattern.c#L749). There's an example of this exact lifecycle pattern in the Cairo source code [here](https://github.molgen.mpg.de/git-mirror/cairo/blob/747cab741cf63f30aa6bc2b787fc13e7e7a2b859/src/cairo-default-context.c#L238-L255). Make sense?

Reply:
i thought that since the pattern is alive and to be used you have to keep the surface.
I get this feeling from the sentence 'the pattern owns the surface'  but maybe i m wrong. I'll try to delete everything, if it still draws is ok.